File: javascript/tree/iterativeDfs.js
/* Definition for a binary tree node:
    class TreeNode {
        constructor(val, left, right) {
             this.val = val;
             this.left = left;
             this.right = right;
        }
    }
*/

// Time and space: O(n)
function inorder(root) {
    let stack = [];
    let curr = root;

    while (curr != null || stack.length != 0) {
        while (curr != null) {
            stack.push(curr);
            curr = curr.left;
        }

        curr = stack.pop();
        console.log(curr.val);
        curr = curr.right;
    }  
}

// Time and space: O(n)
function preorder(root) {
    let stack = [];
    let curr = root;

    while (curr != null || stack.length != 0) {
        if (curr != null) {
            console.log(curr.val);
            if (curr.right != null) {
                stack.push(curr.right);
            }
            curr = curr.left;
        } else {
            curr = stack.pop();
        }
    }
}

// Time and space: O(n)
function postorder(root) {
    let stack = [root];
    let visit = [false];

    while (stack.length != 0) {
        let curr = stack.pop();
        let visited = visit.pop();
        if (curr != null) {
            if (visited) {
                console.log(curr.val);
            } else {
                stack.push(curr);
                visit.push(true);
                stack.push(curr.right);
                visit.push(false);
                stack.push(curr.left);
                visit.push(false);
            }
        }
    }
}

File: python/tree/iterativeDfs.py
# TEST HERE:
https://leetcode.com/problems/binary-tree-inorder-traversal/
https://leetcode.com/problems/binary-tree-preorder-traversal/
https://leetcode.com/problems/binary-tree-postorder-traversal/
# DONT COMMIT THIS BLOCK IN UR CODE U JACKASS

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right

# Time and space: O(n)
def inorder(root):
    stack = []
    curr = root

    while curr or stack:
        if curr:
            stack.append(curr)
            curr = curr.left
        else:
            curr = stack.pop()
            print(curr.val)
            curr = curr.right
    return res

# Time and space: O(n)
def preorder():
    stack = []
    curr = root
    while curr or stack:
        if curr:
            print(curr.val)
            if curr.right:
                stack.append(curr.right)
            curr = curr.left
        else:
            curr = stack.pop()
    return res

# Time and space: O(n)
def postorder():
    stack = [root]
    visit = [False]
    while stack:
        curr, visited = stack.pop(), visit.pop()
        if curr:
            if visited:
                print(curr.val)
            else:
                stack.append(curr)
                visit.append(True)
                stack.append(curr.right)
                visit.append(False)
                stack.append(curr.left)
                visit.append(False)
    return res